x/metadata/types/msg.go

```diff
@@ -423,6 +425,46 @@ func (msg MsgAddScopeSpecificationRequest) ValidateBasic() error {
 	return msg.Specification.ValidateBasic()
 }
 
+// ------------------  MsgAddContractSpecRequest  ------------------
+
+// NewMsgAddContractSpecRequest creates a new msg instance
+func NewMsgAddContractSpecRequest() *MsgAddP8EContractSpecRequest {
+	return &MsgAddP8EContractSpecRequest{}
+}
+
+func (msg MsgAddP8EContractSpecRequest) String() string {
+	out, _ := yaml.Marshal(msg)
+	return string(out)
+}
+
+// Route returns the module route
+func (msg MsgAddP8EContractSpecRequest) Route() string {
+	return ModuleName
+}
+
+// Type returns the type name for this msg
+func (msg MsgAddP8EContractSpecRequest) Type() string {
+	return TypeMsgAddP8EContractSpecRequest
+}
+
+// GetSigners returns the address(es) that must sign over msg.GetSignBytes()
+func (msg MsgAddP8EContractSpecRequest) GetSigners() []sdk.AccAddress {
+	return stringsToAccAddresses(msg.Signers)
+}
+
+// GetSignBytes gets the bytes for the message signer to sign on
+func (msg MsgAddP8EContractSpecRequest) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&msg))
+}
+
+// ValidateBasic performs a quick validity check
+func (msg MsgAddP8EContractSpecRequest) ValidateBasic() error {
+	if len(msg.Signers) < 1 {
+		return fmt.Errorf("at least one signer is required")
+	}
```

Comment:
the internal request should be parsed into the new types here and each of those should have their validatebasic called.

Reply:
That seems pretty wasteful to do here. Whenever we have a MsgAddP8EContractSpecRequest we're going to want to convert it. But the pattern we have is to have ValidateBasice() just return an error. So we'd be converting it, checking the converted pieces, then throwing them away for the error only to need to convert them again later.

In my opinion, we'd just want really superficial validation here, then we'd do the conversion work and validate all the converted things later on.

Reply:
... of course, after I say all that, I look at my own code and realize that converting the msg and throwing away the converted pieces is exactly what I was doing myself. So who knows.

Reply:
I was thinking it would be a bit much too, but @iramiller made a good point about gas usage.

Reply:
It is worth keeping in mind that ValidateBasic checks are performed in a different part of the request flow than the message handler ... they are called in the CheckTx step which is ran whenever a transaction is added to the mem pool.. The purpose of this check is to prevent invalid transaction from hitting the inside of a block ...

The resources used if you let this go into DeliverTx and it gets written into a block as an invalid transaction are infinitely higher than just some extra cpu cycles for processing that occurs here.  If it goes to the next step it gets recorded on the chain forever.